File: pieces/rest_api_postgres/rust/src/models/book.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Debug, sqlx::Type)]
#[sqlx(rename_all = "snake_case")]
pub enum BookStatus {
    Read,
    CurrentlyReading,
    WantToRead,
}

#[derive(Serialize, Deserialize, Debug, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Book {
    pub title: String,
    pub author_first_name: String,
    pub author_last_name: String,
    pub book_status: BookStatus,
    pub date_added: DateTime<Utc>,
    pub date_read: Option<DateTime<Utc>>,
    pub rating: Option<i16>,
}

#[derive(Serialize, Deserialize, Debug, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct BookInDb {
    pub id: i32,
    pub title: String,
    pub author_first_name: String,
    pub author_last_name: String,
    pub book_status: BookStatus,
    pub date_added: DateTime<Utc>,
    pub date_read: Option<DateTime<Utc>>,
    pub rating: Option<i16>,
}

pub fn is_valid_rating(rating: &Option<i16>) -> bool {
    if let Some(r) = rating {
        if r > &5 {
            return false;
        }
    }

    true
}

File: pieces/rest_api_postgres/python/app/models/book.py
from datetime import datetime
from enum import Enum
from typing import Union

from camel_converter.pydantic_base import CamelBase
from pydantic import field_validator


class BookStatus(str, Enum):
    READ = "read"
    CURRENTLY_READING = "currently_reading"
    WANT_TO_READ = "want_to_read"


class Book(CamelBase):
    title: str
    author_first_name: str
    author_last_name: str
    book_status: BookStatus
    date_added: datetime
    date_read: Union[datetime, None] = None
    rating: Union[int, None] = None

    @field_validator("rating")
    @classmethod
    def validate_rating(cls, v: Union[int, None]) -> Union[int, None]:
        if not v:
            return v

        if v < 0 or v > 5:
            raise ValueError("Rating must be between 0 and 5")

        return v


class BookInDb(Book):
    id: int
